packages/ra-ui-materialui/src/button/SaveButton.tsx

```diff
@@ -62,100 +59,56 @@ import { FormRenderProps } from 'react-final-form';
-                console.warn(
-                    '<SaveButton onSave> prop is deprecated, use the onSuccess prop instead.'
-                );
-                if (!formContext || !formContext.setOnSave) {
-                    console.warn(
-                        'Using <SaveButton> outside a FormContext is deprecated.'
-                    );
-                }
-            }
-            if (formContext && formContext.setOnSave) {
-                formContext.setOnSave(onSave);
-            }
-        } else {
-            if (
-                process.env.NODE_ENV !== 'production' &&
-                (!formContext || !formContext.setOnSave)
-            ) {
-                console.warn(
-                    'Using <SaveButton> outside a FormContext is deprecated.'
-                );
-            }
+    const getFormValues = useGetFormValues();
+    const saveContext = useSaveContext();
+    const hasSideEffects = !!onSuccess || !!onFailure || !!transform;
+    const type = !submitOnEnter || hasSideEffects ? 'button' : 'submit';
 
-            if (formContext && formContext.setOnSave) {
-                // we reset to the Form default save function
-                formContext.setOnSave();
-            }
-        }
-        if (onSuccess) {
-            setOnSuccess(onSuccess);
-        }
-        if (onFailure) {
-            setOnFailure(onFailure);
-        }
-        if (transform) {
-            setTransform(transform);
+    const handleClick: MouseEventHandler<HTMLButtonElement> = event => {
+        if (onClick) {
+            onClick(event);
         }
-        if (saving) {
-            // prevent double submission
-            event.preventDefault();
-        } else {
-            if (invalid) {
-                notify('ra.message.invalid_form', { type: 'warning' });
-            }
-            // always submit form explicitly regardless of button type
-            if (event) {
-                event.preventDefault();
-            }
-            handleSubmitWithRedirect(redirect);
+        console.log('defaultPrevented', event.defaultPrevented);
+        if (event.defaultPrevented) {
+            return;
         }
 
-        if (typeof onClick === 'function') {
-            onClick(event);
+        if (hasSideEffects) {
+            console.log({ hasSideEffects });
+            event.preventDefault();
+            const values = getFormValues();
+            saveContext?.save(values, {
+                onSuccess,
+                onFailure,
+                transform,
+            });
         }
     };
 
-    const type = submitOnEnter ? 'submit' : 'button';
     const displayedLabel = label && translate(label, { _: label });
     return (
         <StyledButton
             className={classnames(SaveButtonClasses.button, className)}
             variant={variant}
             type={type}
-            onClick={handleClick}
-            // TODO: find a way to display the loading state (LoadingButton from mui Lab?)
```

Comment:
I'd keep this comment: LoadingButton is a nice addition

Reply:
https://mui.com/components/buttons/#loading-button